doc/ps_design.md

```diff
@@ -0,0 +1,382 @@
+
+Each worker has a local copy of the model variables. After the master relaunches a PS, the PS pod can recover model variables from workers. For embedding vectors, PS must create replicas to support fault tolerance. For each PS *P(i)*, it will store *M* replicas in the following *M* PSs from *P(i+1 % N)* to *P(i+M % N)*. The relaunched PS can recover embedding vectors from one of its replicas. If there are more than *M* continuously-indexed PSs failing, at least one PS fails with all of its replica PSs. The ElasticDL job has to recover from a recent checkpoint.
+
+
+
+## PServer
+
+
+### Compoments
+
+
+![pserver](./images/pserver.png)
+
+
+PServer contains three main compoments:
+
+- KVStore
+- Gradient Queue
+- Optimizer
+
+1. The worker initializes a model, and pushes parameters to KVStore.
+2. Before each step of training, the worker pulls the latest model from KVStore.
+3. After a round of forward/backward computation, the worker pushes gradients to the Gradient Queue waiting for processing.
+4. Optimizer gets a gradient from the Gradient Queue.
+5. Then, Optimizer looks up the corresponding parameter from KVStore.
+6. Optimizer applies gradients to parameters, and updates parameter back to KVStore.
+
+
+### Tensor Data Structure
+
+To support data communication between pods, we introduce a `Tensor` proto message:
+
+```proto
+message Tensor {
+    enum DataType {
+        BOOL = 0;
+        INT16 = 1;
+        INT32 = 2;
+        INT64 = 3;
+        FP16 = 4;
+        FP32 = 5;
+        FP64 = 6;
+    }
+    string name = 1;
+    DataType data_type = 2;
+    repeated int64 dims = 3;
+    bytes content = 4;
+    repeated int64 indices = 5;
+}
+```
+
+Correspondingly, we have a `Tensor` Python class.
+
+
+```python
+class Tensor(object):
+    def __init__(self, name=None, value=None, indices=None):
+        self.name = name
+        self.value = value
+        self.indices = indices
+```
+
+There are also some helper functions:
+
+```python
+def serialize_to_pb(tensor, pb):
+    pass
+
+def deserialize_from_pb(pb, tensor):
+    pass
+
+def convert_to_tf_variable(tensor):
+    pass
+
+def convert_to_tf_tensor(tensor):
+    pass
+```
+
+### KVStore
+
+For a common model variable, we use save it as a `tf.Variable` in Parameter DB. For the embedding table, we introduce a customized data structure.
+
+
+```python
+class EmbeddingTable(object):
+    def __init__(self, name):
+        self.name = name
+        self.vectors = {}
+        
+    def get(self, indices):
+        pass
+        
+    def set(self, indices, value):
+        pass        
+```
```

Comment:
Question 2 EmbeddingTable structure or EmbeddingTableDB structure

KVStore has an embedding_table_db object.
```
class EmbeddingTableDB(object):
    def get_embedding_vector(self, layer_name, slot_name, indices):
        pass

    def set_embedding_vector(self, layer_name, slot_name, indices, value):
        pass

``` 


We should put forward this data structure,  the interface is determined, but we could consider better implementation, even implemented it in C++.

Reply:
Just define the functions, not a class. So the function implementation can be changed later.

Reply:
The functions will manipulate the data. Now, we use a simple dictionary to store all the embedding related param/slot variable etc. 
Function combines data is actually a class. I mean making the function as an interface, the definition is of the function keeps the same.

Reply:
The interface should be changed like this:

```python
class EmbeddingTableDB(object):
    def get_embedding_vector(self, layer_name, slot_name, indices, embedding_meta_info):
        pass
```
The embedding table is lazily initialized.

Reply:
Who will call `get_embedding_vector`? GRPC service? If so, PS should already have `embedding_meta_info`, and we should not send it in every GRPC calls.

Reply:
Not GRPC service. It's used by Optimizer. The optimizer will get embedding vectors from the DB to do updating.